### PR TITLE
Notification Feature

### PR DESCRIPTION
**Pull request by Riken-Shah**
Added notification workflow.

Changes I have made:
- Added the Alert Model
- Integrated Alert Model with the existing API endpoint to create an alert
- Created a new API endpoint to send new alerts
- Created a new API endpoint to mark alerts seen
- Created a hamburger menu for navbar (for smaller width devices)
- Add Notification UI in the navbar

### CODE REVIEW

File: dscblog/paths.py [hunk 1]
```diff
@@ -451,6 +454,7 @@ def blog_comment(request):
             else:
                 comment = request.user.comment(
                     blog=b, text=request.POST['text'].strip(), reference=ref)
+                Alert.create_alert(ref_user=request.user, type=Alert.COMMENT_REPLY if ref else Alert.COMMENT, blog=b)
```
Comment (asrient) [inline dscblog/paths.py:457]: The user getting the alert for COMMENT_REPLY should be the ref comment's user (the person who the user replied to) not the blog author. This is causing some bugs like if i comment on someone's blog and the author replies me back im not getting the alert (as im not the author)

File: dscblog/pages/header.html [hunk 2]
```diff
@@ -7,14 +7,41 @@
...
         <a class="hd_opt btn btn-bordered" href="/profile">Profile</a>
         {% else %}
         <a class="hd_opt btn btn-bordered" href="/accounts/login">Log In</a>
         {% endif %}
+        {% endif %}
+            </div>
+        </div>
+        {% if header.is_loggedin %}
+        <div class="notification-wrapper" id="notification-wrapper">
+            <a class="hd_opt btn btn-bordered" href="#" id="notification-bell">
```
Comment (asrient) [inline dscblog/pages/header.html]: remove the `href="#"` its interfering with navigation history in an unwanted way

File: dscblog/static/styles/header.css [hunk 3]
```diff
@@ -84,8 +176,88 @@
...
+    display: flex;
+    top: 0;
+    left: 0;
+    width: 100vw;
+    height: 100vh;
+    justify-content: center;
+    align-content: center;
+    background: var(--tertiary);
+    z-index: 3;
+    transition: transform 1s linear;
```
Comment (asrient) [inline dscblog/static/styles/header.css]: make it 0.2s

File: dscblog/paths.py (hunk 1 above)
Comment (asrient) [inline dscblog/paths.py]: The user getting the alert for COMMENT_REPLY should be the ref comment's user (the person who the user replied to) not the blog author. This is causing some bugs like if i comment on someone's blog and the author replies me back im not getting the alert (as im not the author)

File: dscblog/pages/header.html [hunk 4]
```diff
@@ -44,12 +72,119 @@
...
+    });
+
+    // Hide Notification Function
+    const hideNotifications = () => {
+        const $notificationWrapper = $('.notification-wrapper');
+        $notificationWrapper.removeClass('active');
+        let alert_ids = [];
+        $('.notification-wrapper .alert').each((i, alert) => alert_ids = [...alert_ids, ...alert.id.split(',')]);
+        setAlertsSeen(alert_ids);
+        fetchNewAlerts();
```
Comment (asrient) [inline dscblog/pages/header.html]: Call `fetchNewAlerts()` after `setAlertsSeen()` is done completely (use a callback). Cuz these are async functions, if `fetchNewAlerts()` gets done before `setAlertsSeen()` finishes, the notif bubble will still remain.

File: dscblog/models.py [hunk 5]
```diff
@@ -859,3 +859,140 @@ def pickOne(cls):
...
+        grouped_alerts = []
+        blogs = {}
+        info = {}
+        for alert in alerts:
+            if alert.type == cls.FOLLOW:
+                grouped_alerts.append({'alert': f'<a href="/@{alert.ref_user.username}">{alert.ref_user}</a> started '
+                                                f'following you.', 'timestamp': alert.timestamp, 'id': alert.pk})
+            elif alert.type == cls.NEW_BLOG:
+                grouped_alerts.append(
+                    {'alert': f'<a href="/@{alert.ref_user.username}">{alert.ref_user}</a> posted a new blog on <a href="{alert.blog.get_url()}">{alert.blog.title}</a>.', 'timestamp': alert.timestamp, 'id': alert.pk})
```
Comment (asrient) [inline dscblog/models.py]: make it `user1 posted a new blog "[blog title]"`

File: dscblog/models.py [hunk 6]
```diff
@@ -859,3 +859,140 @@ def pickOne(cls):
...
+            for alert_type in blog.keys():
+                if blog[alert_type]:
+                    if count == len(blog) - 1 and is_added:
+                        msg += 'and '
+                    msg += f'{blog[alert_type]} {cls.map_alert_type(alert_type)}, '
+                    is_added = True
+                count += 1
+            msg = msg[:-2]
+            if is_added:
+                msg += f" on <a href={info[key]['url']}>{info[key]['name']}</a> blog"
```
Comment (asrient) [inline dscblog/models.py]: make it `on "[blog title]"`

File: dscblog/models.py [hunk 7]
```diff
@@ -859,3 +859,140 @@ def pickOne(cls):
...
+    FOLLOW = 'FL'
+    COMMENT = 'CM'
+    COMMENT_REPLY = 'CR'
+    REACTION = 'RC'
+    NEW_BLOG = 'NB'
+    TYPES = [
+        (FOLLOW, 'Follower'),
+        (COMMENT, 'Comment'),
+        (COMMENT_REPLY, 'Comment Reply'),
+        (REACTION, 'Reaction'),
```
Comment (asrient) [inline dscblog/models.py]: use small case for the description eg: `(COMMENT, 'comment')`

File: dscblog/models.py [hunk 8]
```diff
@@ -859,3 +859,140 @@ def pickOne(cls):
...
+            grouped_alerts.append({'alert': msg, 'timestamp': info[key]['timestamp'], 'id': info[key]['id']})
+
+        return sorted(grouped_alerts, key=lambda x: x['timestamp'], reverse=True)
+
+    @classmethod
+    def map_alert_type(cls, type):
+        for alert_type in cls.TYPES:
+            if alert_type[0] == type:
+                return alert_type[1]
+        return type
```
Comment (asrient) [inline dscblog/models.py]: make it work for plural terms as well, like `comments`, `reactions`

File: dscblog/static/styles/header.css [hunk 9]
```diff
@@ -46,12 +46,104 @@
...
+
+.notification-wrapper .notification{
+  width: 400px;
+  max-height: 250px;
+  position: absolute;
+  display: flex;
+  flex-direction: column;
+  top: 50px;
+  right: 10px;
+  overflow-y: scroll;
```
Comment (asrient) [inline dscblog/static/styles/header.css]: make it `auto`

File: dscblog/static/styles/header.css [hunk 10]
```diff
@@ -46,12 +46,104 @@
...
+  background: rgba(219, 219, 219, 1);
+  border: 1px solid rgba(0,0,0,.15);
+  box-shadow: 0 6px 12px rgba(0,0,0,.175);
+  z-index: 0;
+  opacity: 0;
+  transition: all .3s ease-in;
+  -ms-overflow-style: none; /* IE and Edge */
+  scrollbar-width: none; /* Firefox */
+
```
Comment (asrient) [inline dscblog/static/styles/header.css]: add: 
border-radius: 0.2rem;
padding: 0.5rem 0.7rem;

File: dscblog/static/styles/header.css [hunk 11]
```diff
@@ -46,12 +46,104 @@
...
+  max-height: 250px;
+  position: absolute;
+  display: flex;
+  flex-direction: column;
+  top: 50px;
+  right: 10px;
+  overflow-y: scroll;
+  background: rgba(219, 219, 219, 1);
+  border: 1px solid rgba(0,0,0,.15);
+  box-shadow: 0 6px 12px rgba(0,0,0,.175);
```
Comment (asrient) [inline dscblog/static/styles/header.css]: for dark mode:
`box-shadow: 4px 5px 0px rgb(158 158 158);`
for light mode, change the shadow color to dark (more consistent with the theme)

File: dscblog/static/styles/header.css [hunk 12]
```diff
@@ -46,12 +46,104 @@
...
+.notification-wrapper .notification{
+  width: 400px;
+  max-height: 250px;
+  position: absolute;
+  display: flex;
+  flex-direction: column;
+  top: 50px;
+  right: 10px;
+  overflow-y: scroll;
+  background: rgba(219, 219, 219, 1);
```
Comment (asrient) [inline dscblog/static/styles/header.css]: rgb(244 244 244)

File: dscblog/static/styles/header.css [hunk 13]
```diff
@@ -46,12 +46,104 @@
...
+  transition: all .3s ease-in;
+  -ms-overflow-style: none; /* IE and Edge */
+  scrollbar-width: none; /* Firefox */
+
+}
+.notification-wrapper .notification::-webkit-scrollbar {
+  display: none;
+}
+[data-theme="dark"] .notification-wrapper .notification{
+  background: rgb(55, 55, 55);
```
Comment (asrient) [inline dscblog/static/styles/header.css]: rgb(38 38 38)

File: dscblog/static/styles/header.css [hunk 14]
```diff
@@ -46,12 +46,104 @@
...
+  width: 400px;
+  max-height: 250px;
+  position: absolute;
+  display: flex;
+  flex-direction: column;
+  top: 50px;
+  right: 10px;
+  overflow-y: scroll;
+  background: rgba(219, 219, 219, 1);
+  border: 1px solid rgba(0,0,0,.15);
```
Comment (asrient) [inline dscblog/static/styles/header.css]: For dark mode:
`border: 1px solid rgb(255 255 255 / 95%);`
For light mode:
`border: 1px solid rgb(0 0 0 / 95%);`

File: dscblog/pages/header.html [hunk 15]
```diff
@@ -44,12 +72,119 @@
...
+        $('#old-alerts div').remove();
+
+        if (res.new_alerts && res.new_alerts.length) {
+            $('#total_new_alerts').show();
+            const totalNewAlerts = res.new_alerts.length;
+            $('#total_new_alerts')[0].innerHTML = totalNewAlerts < 10 ? totalNewAlerts : '9+';
+            res.new_alerts.forEach((alert) => $('#new-alerts').append(`<div class="alert" id=${alert.id}><div>${alert.alert}</div><span class="timespan">${humanized_time_span(alert.timestamp)}</span></div>`));
+        } else {
+            $('#total_new_alerts').hide();
+            $('#new-alerts').append('<div class="alert">Sorry, No new alerts found.</div>');
```
Comment (asrient) [inline dscblog/pages/header.html]: instead of this, just hide the new alerts section

File: dscblog/pages/header.html [hunk 16]
```diff
@@ -44,12 +72,119 @@
...
+        }
+
+        if (res.old_alerts && res.old_alerts.length) {
+            res.old_alerts.forEach((alert) => $('#old-alerts').append(`<div class="alert" id=${alert.id}><div>${alert.alert}</div><span class="timespan">${humanized_time_span(alert.timestamp)}</span></div>`));
+        } else {
+            $('#old-alerts').append('<div class="alert">Sorry, No old alerts found.</div>');
```
Comment (asrient) [inline dscblog/pages/header.html]: instead, hide the old alerts section

File: dscblog/models.py [hunk 17]
```diff
@@ -859,3 +859,140 @@ def pickOne(cls):
...
+
+        for key in blogs.keys():
+            msg = f'You have '
+            count = 0
+            is_added = False
+            blog = blogs[key]
+            for alert_type in blog.keys():
+                if blog[alert_type]:
+                    if count == len(blog) - 1 and is_added:
+                        msg += 'and '
```
Comment (asrient) [inline dscblog/models.py:984]: add the word `new` to msg, like `You have 1 new comment on...`

File: dscblog/static/styles/header.css [hunk 18]
```diff
@@ -46,12 +46,104 @@
...
+  box-shadow: 0 6px 12px rgba(0,0,0,.175);
+  z-index: 0;
+  opacity: 0;
+  transition: all .3s ease-in;
+  -ms-overflow-style: none; /* IE and Edge */
+  scrollbar-width: none; /* Firefox */
+
+}
+.notification-wrapper .notification::-webkit-scrollbar {
+  display: none;
```
Comment (asrient) [inline dscblog/static/styles/header.css]: make the scrollbar hide only on mobile, on desktop use a custom sleek scrollbar

File: dscblog/static/styles/header.css [hunk 19]
```diff
@@ -46,12 +46,104 @@
...
+
+}
+.notification-wrapper .notification .alerts-wrapper .alert{
+  margin: 10px 0;
+  position: relative;
+  display: flex;
+  flex-direction: column;
+  font-size: .9rem;
+  padding-left: 5px;
+  border-bottom: solid .5px #959090;
```
Comment (asrient) [inline dscblog/static/styles/header.css]: make the border color lighter..
for dark mode: #474747
light mode: #c8c8c8

File: dscblog/static/styles/header.css [hunk 20]
```diff
@@ -46,12 +46,104 @@
...
+.notification-wrapper .notification .alerts-wrapper .type {
+  font-weight: 700;
+  border-bottom: solid 1px;
+  width: 100%;
+  display: block;
+  padding-left: 5px;
+  padding-bottom: 5px;
+
+}
+.notification-wrapper .notification .alerts-wrapper .alert{
```
Comment (asrient) [inline dscblog/static/styles/header.css:124]: remove the border-bottom for the last item:
```css
.notification-wrapper .notification .alerts-wrapper .alert:last-child {
    border-bottom: none;
}
```

File: dscblog/static/styles/header.css [hunk 21]
```diff
@@ -46,12 +46,104 @@
...
+  margin: 10px 0;
+  position: relative;
+  display: flex;
+  flex-direction: column;
+  font-size: .9rem;
```
Comment (asrient) [inline dscblog/static/styles/header.css]: font-size: .85rem;
color: var(--grey); //For older alerts
color: rgb(186 186 186); //For new alerts (dark m)
color: rgb(59 59 59); //For new alerts (light m)

File: dscblog/static/styles/header.css [hunk 22]
```diff
@@ -46,12 +46,104 @@
...
+.notification-wrapper .notification .alerts-wrapper .alert a{
+  color: #1184ff;
+}
+.notification-wrapper .notification .alerts-wrapper .timespan{
+  font-size: 11px;
+  padding: 5px 0;
+  color: #827e7e;
+  font-weight: 100;
+}
+
```
Comment (asrient) [inline dscblog/static/styles/header.css:152]: Add:
```css
.notification-wrapper .notification .alerts-wrapper .alert a {
    color: var(--secondary);
    font-weight: 600;
}
```
for older alerts, change the color to less contrasting.

File: dscblog/models.py [hunk 23]
```diff
@@ -859,3 +859,140 @@ def pickOne(cls):
...
+        return True
+
+    @classmethod
+    def group_alerts(cls, alerts):
+        grouped_alerts = []
+        blogs = {}
+        info = {}
+        for alert in alerts:
+            if alert.type == cls.FOLLOW:
+                grouped_alerts.append({'alert': f'<a href="/@{alert.ref_user.username}">{alert.ref_user}</a> started '
```
Comment (asrient) [inline dscblog/models.py]: use the actual `name` of the user instead of their `username` if avaiable

File: dscblog/pages/header.html [hunk 24]
```diff
@@ -7,14 +7,41 @@
...
+        </div>
+        {% if header.is_loggedin %}
+        <div class="notification-wrapper" id="notification-wrapper">
+            <a class="hd_opt btn btn-bordered" href="#" id="notification-bell">
+                <img src="/static/icons/notification_dark.png" alt="Bell Icon"/>
+                <span class="alert_number" id="total_new_alerts">0</span>
+            </a>
+            <div class="notification">
+                <div class="alerts-wrapper" id="new-alerts">
+                    <span class="type">New Alerts</span>
```
Comment (asrient) [inline dscblog/pages/header.html]: rename it `New`

File: dscblog/pages/header.html [hunk 25]
```diff
@@ -7,14 +7,41 @@
...
+                </div>
+                <div class="alerts-wrapper" id="old-alerts">
+                    <span class="type">Old Alerts</span>
```
Comment (asrient) [inline dscblog/pages/header.html]: rename it `Earlier`

**Review by asrient — CHANGES_REQUESTED** [review]
Hi!
I have requested some changes, pls update ur PR accordingly.

File: dscblog/models.py (hunk 7 above)
Comment (Riken-Shah) [reply dscblog/models.py]: Ok

File: dscblog/pages/header.html (hunk 15 above)
Comment (Riken-Shah) [reply dscblog/pages/header.html]: What would happen If the user has no alerts?

Comment (asrient) [reply dscblog/pages/header.html]: Put a text positioned centre of the modal saying something like "No new notifications" or something cool.

**Review by asrient — APPROVED** [review]

**Pull request merged**